File: android/app/src/main/java/org/openbot/env/ConnectionManager.java
package org.openbot.env;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

public class ConnectionManager {
  private static final String TAG = "ConnectionManager";
  private static ConnectionManager _connectionManager;
  private static Context _context;
  private ILocalConnection connection;

  private final ILocalConnection networkConnection = new NetworkServiceConnection();
  private final ILocalConnection nearbyConnection = new NearbyConnection();

  private ConnectionManager() {
    if (_connectionManager != null) {
      throw new RuntimeException(
          "Use getInstance() method to get the single instance of this class.");
    }
  }

  public static ConnectionManager getInstance(Context context) {
    _context = context;
    if (_connectionManager == null) {

      synchronized (ConnectionManager.class) {
        if (_connectionManager == null) _connectionManager = new ConnectionManager();
      }
    }

    return _connectionManager;
  }

  ILocalConnection getConnection() {
    if (connection != null) {
      return connection;
    }

    if (isConnectedViaWifi()) {
      connection = networkConnection;
    } else {
      connection = nearbyConnection;
    }

    return connection;
  }

  private boolean isConnectedViaWifi() {
    ConnectivityManager connectivityManager =
        (ConnectivityManager) _context.getSystemService(Context.CONNECTIVITY_SERVICE);
    NetworkInfo mWifi = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
    return mWifi.isConnected();
  }
}

File: android/controller/src/main/java/org/openbot/controller/ConnectionManager.kt
package org.openbot.controller

import android.annotation.SuppressLint
import android.content.Context
import android.content.Context.WIFI_SERVICE
import android.net.wifi.WifiManager

@SuppressLint("StaticFieldLeak")
object ConnectionManager {

    private val connection: ILocalConnection? = null
    private const val TAG = "ConnectionManager"
    @SuppressLint("StaticFieldLeak")
    private lateinit var context: Context

    fun init (context: Context) {
        this.context = context;
    }

    fun getConnection(): ILocalConnection {
        val connected = isConnectedViaWifi(context)
        return if (connected) NetworkServiceConnection else NearbyConnection
    }

    private fun isConnectedViaWifi(context: Context): Boolean {

        val wifiManager = context.getSystemService(WIFI_SERVICE) as WifiManager?
        val info = wifiManager!!.connectionInfo ?: return false
        val networkId = info.networkId

        return networkId > 0
    }
}
